<a href="https://colab.research.google.com/github/zhuoyingcai/NYC-Housing-Prices-Analysis/blob/master/NYC_Housing_Prices_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Section

In [0]:
import os

# Data
import numpy as np
import pandas as pd

# Plot
import matplotlib.pyplot as plt


# Path Functions

In [0]:
def ensure_dir(path_to_dir):
  try:
      os.makedirs(path_to_dir)
  except FileExistsError:
      # directory already exists
      pass

# Config

In [3]:
# path config
annualizedSalesUpdate_data_dir = '/content/drive/My Drive/Colab Notebooks/dataScience/annualizedSalesUpdate'
data_dir = '/content/drive/My Drive/Colab Notebooks/dataScience/data'

# mount google drive
from google.colab import drive
drive.mount('/content/drive')

# make sure dir exist
ensure_dir(data_dir)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# 1. Load and summarize data

In [0]:
# define dataframe
df = pd.DataFrame()

In [0]:
years = [2017, 2018]
boroughs = ['bronx', 'brooklyn', 'manhattan', 'queens', 'statenisland']

# load data
for year in years:
  for borough in boroughs:
    path = annualizedSalesUpdate_data_dir + '/%s_%s.xls' % (year, borough)
    data = pd.read_excel(path, header=4)
    df = df.append(data, sort=False)

We first visually inspect the data to see if there is any missing data or incorrect data types. Also, we may want to edit data entry.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 170407 entries, 0 to 8385
Data columns (total 21 columns):
BOROUGH                           170407 non-null int64
NEIGHBORHOOD                      170407 non-null object
BUILDING CLASS CATEGORY           170407 non-null object
TAX CLASS AT PRESENT              170349 non-null object
BLOCK                             170407 non-null int64
LOT                               170407 non-null int64
EASE-MENT                         88660 non-null object
BUILDING CLASS AT PRESENT         170349 non-null object
ADDRESS                           170407 non-null object
APARTMENT NUMBER                  106536 non-null object
ZIP CODE                          170393 non-null float64
RESIDENTIAL UNITS                 170349 non-null float64
COMMERCIAL UNITS                  170349 non-null float64
TOTAL UNITS                       170349 non-null float64
LAND SQUARE FEET                  170346 non-null float64
GROSS SQUARE FEET                 1

In [7]:
df.head(10)

BOROUGH NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
0        2     BATHGATE  01 ONE FAMILY DWELLINGS                       
1        2     BATHGATE  01 ONE FAMILY DWELLINGS                       
2        2     BATHGATE  01 ONE FAMILY DWELLINGS                       
3        2     BATHGATE  01 ONE FAMILY DWELLINGS                       
4        2     BATHGATE  01 ONE FAMILY DWELLINGS                       
5        2     BATHGATE  01 ONE FAMILY DWELLINGS                       
6        2     BATHGATE  01 ONE FAMILY DWELLINGS                       
7        2     BATHGATE  01 ONE FAMILY DWELLINGS                       
8        2     BATHGATE  01 ONE FAMILY DWELLINGS                       
9        2     BATHGATE  01 ONE FAMILY DWELLINGS                       

  TAX CLASS AT PRESENT  BLOCK  LOT EASE-MENT BUILDING CLASS AT PRESENT  \
0                    1   3028   25                                  A5   
1                    1   3030   55                                  A5   
2                    1   3030   56                                  A1   
3                    1   3030   56                                  A1   
4                    1   3030   65                                  A1   
5                    1   3030   70                                  A1   
6                    1   3036   13                                  A1   
7                    1   3037   42                                  A1   
8                    1   3037  101                                  S0   
9                    1   3043   55                                  A1   

                 ADDRESS APARTMENT NUMBER  ...  RESIDENTIAL UNITS  \
0    412 EAST 179 STREET                   ...                1.0   
1  410 EAST 182ND STREET                   ...                1.0   
2    412 EAST 182 STREET                   ...                1.0   
3    412 EAST 182 STREET                   ...                1.0   
4       4455 PARK AVENUE                   ...                1.0   
5       4445 PARK AVENUE                   ...                1.0   
6       4348 PARK AVENUE                   ...                1.0   
7       4428 PARK AVENUE                   ...                1.0   
8    443 EAST 180 STREET                   ...                1.0   
9   1948 BATHGATE AVENUE                   ...                1.0   

   COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET  \
0               0.0          1.0            1842.0             2048.0   
1               0.0          1.0            1330.0             1460.0   
2               0.0          1.0            1306.0             1440.0   
3               0.0          1.0            1306.0             1440.0   
4               0.0          1.0            1622.0             1587.0   
5               0.0          1.0            1694.0             1497.0   
6               0.0          1.0            3525.0             1764.0   
7               0.0          1.0            3525.0             1340.0   
8               2.0          3.0            1293.0             2820.0   
9               0.0          1.0            2356.0             2047.0   

   YEAR BUILT  TAX CLASS AT TIME OF SALE  BUILDING CLASS AT TIME OF SALE  \
0      1901.0                          1                              A5   
1      1899.0                          1                              A5   
2      1899.0                          1                              A1   
3      1899.0                          1                              A1   
4      1899.0                          1                              A1   
5      1899.0                          1                              A1   
6      1899.0                          1                              A1   
7      1899.0                          1                              A1   
8      1952.0                          1                              S0   
9      1901.0                          1                              A

In [8]:
df.tail()

BOROUGH NEIGHBORHOOD     BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  \
8381        5      WOODROW  05 TAX CLASS 1 VACANT LAND                   1B   
8382        5      WOODROW  05 TAX CLASS 1 VACANT LAND                    1   
8383        5      WOODROW  05 TAX CLASS 1 VACANT LAND                    1   
8384        5      WOODROW  05 TAX CLASS 1 VACANT LAND                    1   
8385        5      WOODROW  05 TAX CLASS 1 VACANT LAND                   1B   

      BLOCK  LOT EASE-MENT BUILDING CLASS AT PRESENT            ADDRESS  \
8381   7094   61       NaN                        V0  23 WILTSHIRE LANE   
8382   7267  289       NaN                        A5      41 JADE COURT   
8383   7267  298       NaN                        A5      11 JADE COURT   
8384   7267  299       NaN                        A5       9 JADE COURT   
8385   7346    1       NaN                        V0      GAYNOR STREET   

     APARTMENT NUMBER  ...  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  \
8381              NaN  ...                0.0               0.0          0.0   
8382              NaN  ...                1.0               0.0          1.0   
8383              NaN  ...                1.0               0.0          1.0   
8384              NaN  ...                1.0               0.0          1.0   
8385              NaN  ...                0.0               0.0          0.0   

      LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  \
8381            4427.0                0.0      2018.0   
8382            2955.0             2612.0      2018.0   
8383            4108.0             2612.0      2018.0   
8384           11019.0             2612.0      2018.0   
8385           42016.0                0.0         0.0   

      TAX CLASS AT TIME OF SALE  BUILDING CLASS AT TIME OF SALE SALE PRICE  \
8381                          1                              V0     793216   
8382                          1                              V0     671026   
8383                          1                              V0     610950   
8384                          1                              V0     672043   
8385                          1                              V0          0   

      SALE DATE  
8381 2018-05-25  
8382 2018-10-04  
8383 2018-07-23  
8384 2018-08-07  
8385 2018-09-21  

[5 rows x 21 columns]

# 2. Data Cleaning

---

We found two problems:

 1. we see that the dataframe has 170407 entries, but TAX CLASS AT PRESENT, EASE-MENT, BUILDING CLASS AT PRESENT, APARTMENT NUMBER, ZIP CODE, RESIDENTIAL UNITS, COMMERCIAL UNITS, TOTAL UNITS, LAND SQUARE FEET, GROSS SQUARE FEET and YEAR BUILT has less entries.
 
 2. The BROUGH is represented as numerical values rather than categorical.
 
 3. There are some data does not make sense, such as GROSS SQUARE FEET, YEAR BUILT and SALE PRICE can not be 0.


---


 What we are going to do next:
 
 1. rename BOROUGH values to actual borough name.
 
 2. drop columns that we are not going to use.
 
 3. remove raws with 0 SALE PRICE, GROSS SQUARE FEET, LAND SQUARE FEET and YEAR BUILT.

In [0]:
df['BOROUGH'] = df['BOROUGH'].replace({1: 'Manhattan', 2: 'Bronx', 3: 'Brooklyn', 4: 'Queens', 5: 'Staten Island'})

In [0]:
df = df.drop(['BUILDING CLASS CATEGORY', 'TAX CLASS AT PRESENT', 'EASE-MENT', 'BUILDING CLASS AT PRESENT', 'ADDRESS',
         'APARTMENT NUMBER', 'ZIP CODE', 'BUILDING CLASS AT TIME OF SALE' ], axis=1)

In [0]:
df = df[df['SALE PRICE'] != 0]
df = df[df['GROSS SQUARE FEET'] != 0]
df = df[df['LAND SQUARE FEET'] != 0]
df = df[df['YEAR BUILT'] != 0]

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 64388 entries, 1 to 8384
Data columns (total 13 columns):
BOROUGH                      64388 non-null object
NEIGHBORHOOD                 64388 non-null object
BLOCK                        64388 non-null int64
LOT                          64388 non-null int64
RESIDENTIAL UNITS            64346 non-null float64
COMMERCIAL UNITS             64346 non-null float64
TOTAL UNITS                  64346 non-null float64
LAND SQUARE FEET             64345 non-null float64
GROSS SQUARE FEET            64346 non-null float64
YEAR BUILT                   64378 non-null float64
TAX CLASS AT TIME OF SALE    64388 non-null int64
SALE PRICE                   64388 non-null int64
SALE DATE                    64388 non-null datetime64[ns]
dtypes: datetime64[ns](1), float64(6), int64(4), object(2)
memory usage: 6.9+ MB


## 2.1 Outlier
We also want to look at the distribution of our SALE PRICE.

1. We noticed that there are some extremely expensive house located in Manhattan, we would like to look at it and decide whether include them in our project or not.

2. The documentation of this data set explained there will be $0 sales price which means that there was a transfer of ownership without a cash consideration. However, there are some other values indicate there was a transfer of ownership which is not in the consideration of our project.

In [13]:
df.nlargest(10, 'SALE PRICE')

BOROUGH             NEIGHBORHOOD  BLOCK  LOT  RESIDENTIAL UNITS  \
773    Manhattan                  CHELSEA    713    1                0.0   
7347   Manhattan              MIDTOWN CBD   1301    1                0.0   
2572   Manhattan                FINANCIAL     40    3                0.0   
14970  Manhattan  UPPER WEST SIDE (59-79)   1119   61                0.0   
771    Manhattan                  CHELSEA    673    1                0.0   
20902   Brooklyn             SPRING CREEK   4435    1             1092.0   
1755   Manhattan                  FASHION    840   42                0.0   
8602   Manhattan             MIDTOWN WEST   1000    1                0.0   
6795   Manhattan              MIDTOWN CBD   1298   23                0.0   
2570   Manhattan                FINANCIAL     29    1                0.0   

       COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET  \
773               167.0        167.0          165200.0          1100533.0   
7347               35.0         35.0           81336.0          1586886.0   
2572                1.0          1.0           53632.0          1617206.0   
14970               1.0          1.0           15059.0            99769.0   
771                42.0         42.0          136000.0          1129200.0   
20902               0.0       1092.0          358496.0           997720.0   
1755                4.0          4.0           51271.0           697029.0   
8602                4.0          4.0           16066.0           369864.0   
6795               42.0         42.0           35145.0           627691.0   
2570                1.0          1.0           42762.0           993569.0   

       YEAR BUILT  TAX CLASS AT TIME OF SALE  SALE PRICE  SALE DATE  
773        1905.0                          4  2397501899 2018-03-20  
7347       1966.0                          4  2210000000 2017-05-05  
2572       1987.0                          4  1040000000 2017-01-24  
14970      1978.0                          4   980000000 2018-06-15  
771        1912.0                          4   880000000 2018-10-23  
20902      1971.0                          2   869612895 2018-05-04  
1755       1920.0                          4   850000000 2019-02-08  
8602       1909.0                          4   719099999 2018-04-27  
6795       1987.0                          4   701000000 2018-10-30  
2570       1983.0                          4   652000000 2017-05-24

In [14]:
lst = df['SALE PRICE'].unique().tolist()
lst = np.sort(lst, axis = None)
lst[:100]

array([    1,     2,     3,     4,     5,     6,    10,    15,    20,
          60,   100,   125,   200,   210,   396,   416,   500,   700,
         825,  1000,  1006,  1041,  1162,  1500,  1787,  1800,  1895,
        2000,  2352,  2416,  2500,  2567,  3000,  3317,  3500,  3582,
        3750,  3774,  3900,  3937,  4000,  4500,  4700,  4800,  4900,
        4997,  5000,  5041,  5100,  5200,  5300,  5500,  5700,  5750,
        6000,  6081,  6250,  6270,  6300,  6400,  6438,  6500,  6533,
        6562,  6650,  7000,  7500,  7923,  7980,  8000,  8250,  8500,
        9500,  9686,  9828, 10000, 10001, 10050, 10500, 11000, 11500,
       11666, 12000, 12497, 12500, 12678, 12950, 13333, 13864, 13870,
       14583, 14915, 15000, 15238, 15387, 15523, 15567, 16000, 16500,
       16544])

## 2.2 Cut Unreal Price

From the above data, we conclude that:

1. We think the price of the most expensive buildings is reasonable. We decided to keep those data.

2. There are a lot of extremely low sale price. In order to solve this problem, we decided to set a minimum price range for the data.

In [0]:
df = df[(df['SALE PRICE'] > 100000)]

In [16]:
df.describe()

BLOCK           LOT  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
count  61408.000000  61408.000000       61366.000000      61366.000000   
mean    5423.205055    156.263190           2.581234          0.232507   
std     3797.794759    389.148401          15.295299          3.681592   
min        5.000000      1.000000           0.000000          0.000000   
25%     2376.000000     21.000000           1.000000          0.000000   
50%     4821.000000     42.000000           1.000000          0.000000   
75%     7720.000000     74.000000           2.000000          0.000000   
max    16323.000000   5323.000000        1844.000000        436.000000   

        TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET    YEAR BUILT  \
count  61366.000000      6.136500e+04       6.136600e+04  61398.000000   
mean       2.886403      6.458235e+03       3.869931e+03   1944.588407   
std       15.972824      3.400238e+04       2.834616e+04     33.196646   
min        0.000000      1.000000e+00       1.000000e+00   1800.000000   
25%        1.000000      2.000000e+03       1.298000e+03   1920.000000   
50%        1.000000      2.521000e+03       1.792000e+03   1931.000000   
75%        2.000000      4.000000e+03       2.551000e+03   1965.000000   
max     1866.000000      4.228300e+06       3.750565e+06   2018.000000   

       TAX CLASS AT TIME OF SALE    SALE PRICE  
count               61408.000000  6.140800e+04  
mean                    1.269346  1.829425e+06  
std                     0.685776  1.871611e+07  
min                     1.000000  1.000440e+05  
25%                     1.000000  4.800000e+05  
50%                     1.000000  6.830000e+05  
75%                     1.000000  1.000000e+06  
max                     4.000000  2.397502e+09

# 3. Data Visualization